kubeadm upgrade plan: print a component config state table

rosti opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This change enables `kubeadm upgrade plan` to print a state table with information regarding known component config API groups. Most importantly this information includes current and preferred version for each group and an indication if a manual config upgrade is required.

In addition to that, this PR also replaces most of the versioned use of the output API in `kubeadm upgrade plan` with the internal types.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Refs:
- kubernetes/kubeadm#1940
- kubernetes/enhancements#1381

**Special notes for your reviewer**:

This PR is part of the implementation of the new kubeadm component config management scheme KEP (see link below). It also depends on #91980  . Please, review the last 2 commits only!

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/priority important-longterm
/assign @fabriziopandini @neolit123 @ereslibre
/hold

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubeadm: upgrade plan now prints a table indicating the state of known component configs prior to upgrade
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md
```

rosti committed:
kubeadm upgrade plan: component config state table

This change enables kubeadm upgrade plan to print a state table with
information regarding known component config API groups. Most importantly this
information includes current and preferred version for each group and an
indication if a manual user upgrade is required.

Signed-off-by: Rostislav M. Georgiev <rostislavg@vmware.com>

neolit123 commented:
@rosti can you please show example output?

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -75,7 +75,7 @@ func getK8sVersionFromUserInput(flags *applyPlanFlags, args []string, versionIsM
 
 // enforceRequirements verifies that it's okay to upgrade and then returns the variables needed for the rest of the procedure
-func enforceRequirements(flags *applyPlanFlags, dryRun bool, newK8sVersion string) (clientset.Interface, upgrade.VersionGetter, *kubeadmapi.InitConfiguration, error) {
+func enforceRequirements(plan bool, flags *applyPlanFlags, dryRun bool, newK8sVersion string) (clientset.Interface, upgrade.VersionGetter, *kubeadmapi.InitConfiguration, error) {
```
cc @Klaven

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -425,6 +425,9 @@ type ComponentConfig interface {
 	DeepCopy() ComponentConfig
 
+	// Version returns a Kubernetes API version of the component config (v1alpha1, v1beta1, v1, etc.)
+	Version() string
```
i think we might want to have VersionKind() here instead even if our third-party CCs don't have multiple kinds yet.

rosti replied:
I don't want kinds to be visible at all (let alone at this level). What I was thinking, and I'll probably change it on an iteration is to have this as a `GroupVersion`.
The idea behind `ComponentConfig` objects is to have them hide the details around a given `GroupVersion`. This includes what kinds there are, what goes where and how to structure it.
This helps isolate version specific code. Plus, adding new component config versions becomes more easy.

neolit123 replied:
Groups are not supposed to change, but a Group can introduce new Kinds during upgrade to a new Version. not tracking new Kinds means no visibility for this aspect from "plan", correct?

GroupVersion-only tracking would mean that "plan" can say, hey i know there is a new version  for this Group, but i'm not going to tell you what changed in it.

neolit123 replied:
after discussions on this, having `group | version` is fine.

rosti commented:
Sample output:
```
# kubeadm upgrade plan --allow-experimental-upgrades
...
Components that must be upgraded manually after you have upgraded the control plane with 'kubeadm upgrade apply':
COMPONENT   CURRENT                                        AVAILABLE
Kubelet     1 x v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5

Upgrade to the latest experimental version:

COMPONENT            CURRENT                                    AVAILABLE
API Server           v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5
Controller Manager   v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5
Scheduler            v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5
Kube Proxy           v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5
CoreDNS              1.6.5                                      1.6.5
Etcd                 3.4.3                                      3.4.3-0

You can now apply the upgrade by executing the following command:

        kubeadm upgrade apply v1.18.0-alpha.5 --allow-experimental-upgrades

Note: Before you can perform this upgrade, you have to update kubeadm to v1.18.0-alpha.5.

_____________________________________________________________________


The table below shows the current state of component configs as understood by this version of kubeadm.
Configs that have a "yes" mark in the "USER ACTION REQUIRED" column require manual config upgrade or
resetting to kubeadm defaults before a successful upgrade can be performed. The version to manually
upgrade to is denoted in the "PREFERRED VERSION" column.

API GROUP                 CURRENT VERSION   PREFERRED VERSION   USER ACTION REQUIRED
kubeproxy.config.k8s.io   v1alpha1          v1alpha1            no
kubelet.config.k8s.io     v1beta1           v1beta1             no
_____________________________________________________________________
#
```

rosti committed:
kubeadm upgrade plan: Use internal types of the output API

`kubeadm upgrade plan` is using the external (currently `v1alpha1`) types of
the kubeadm output API to collect upgrade plans. This is counter intuitive
since code structure gets bound to the whatever version the output API is at.
In addition to that, the versioned API is used only in the very last stages of
a machine readable output (which is currently not implemented).

Hence, to increase flexibility and keep up with the standard Kubernetes
ecosystem practice, `kubeadm upgrade plan` is migrated to use the internal
types of the output API.

Signed-off-by: Rostislav M. Georgiev <rostislavg@vmware.com>

neolit123 commented:
/milestone v1.19

rosti commented:
/retest

neolit123 commented:
/lgtm
/retest

neolit123 reviewed (commented):
/hold cancel
we have a number of PRs in the queue and it might be quite packed this cycle.
